File: libplugin/src/main/java/com/lemon/plugin/inject/ClassTransform.java
package com.lemon.plugin.inject;

import com.android.build.api.transform.*;
import com.android.build.gradle.TestedExtension;
import com.android.build.gradle.internal.pipeline.TransformManager;
import com.google.common.collect.Sets;
import com.lemon.plugin.inject.tool.asm.AsmInjectTool;
import com.lemon.plugin.inject.tool.javaassist.JavaAssistInjectTool;

import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.io.FileUtils;
import org.gradle.api.Project;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;

class ClassTransform extends Transform {
    private Project mProject;
    private boolean mIsApp;
    private TestedExtension mExtension;

    private IInjectTool mInjectTool;

    ClassTransform(Project project, boolean isApp, TestedExtension extension) {
        mProject = project;
        mIsApp = isApp;
        mExtension = extension;

        mInjectTool = new JavaAssistInjectTool(mProject.getRootDir().getAbsolutePath());
//        mInjectTool = new AsmInjectTool();
    }

    @Override
    public String getName() {
        return "class editor";
    }

    @Override
    public Set<QualifiedContent.ContentType> getInputTypes() {
        return TransformManager.CONTENT_CLASS;
    }

    @Override
    public Set<QualifiedContent.Scope> getScopes() {
        if(mIsApp) {
            return Sets.immutableEnumSet(
                    QualifiedContent.Scope.PROJECT,
                    QualifiedContent.Scope.SUB_PROJECTS,
                    QualifiedContent.Scope.EXTERNAL_LIBRARIES);
        } else {
            return Sets.immutableEnumSet(
                    QualifiedContent.Scope.PROJECT);
        }
    }

    @Override
    public boolean isIncremental() {
        return false;
    }

    @Override
    public void transform(Context context,
                   Collection<TransformInput> inputs,
                   Collection<TransformInput> referencedInputs,
                   TransformOutputProvider outputProvider,
                   boolean isIncremental) throws IOException {
        InjectUtils.println("transform start");
        mProject.getDependencies();
        for (TransformInput input : inputs) {
            for (DirectoryInput directoryInput : input.getDirectoryInputs()) {
                File dest = outputProvider.getContentLocation(
                        directoryInput.getName(),
                        directoryInput.getContentTypes(),
                        directoryInput.getScopes(),
                        Format.DIRECTORY);

                try {
                    // 只有这一句是有用的
                    inject(directoryInput.getFile().getAbsolutePath(), dest.getAbsolutePath());
                } catch (Exception e) {
                    e.printStackTrace();
                }

                FileUtils.copyDirectory(directoryInput.getFile(), dest);
            }

            for (JarInput jarInput : input.getJarInputs()) {
                String jarName = jarInput.getName();
                String md5Name = DigestUtils.md5Hex(jarInput.getFile().getAbsolutePath());
                if (jarName.endsWith(".jar")) {
                    jarName = jarName.substring(0, jarName.length() - 4);
                }
                File dest = outputProvider.getContentLocation(
                        jarName + md5Name,
                        jarInput.getContentTypes(),
                        jarInput.getScopes(),
                        Format.JAR);
                FileUtils.copyFile(jarInput.getFile(), dest);
            }
        }
        InjectUtils.println("transform finish");
    }

    private void inject(String path, String dest) {
        File dir = new File(path);
        if (dir.isDirectory()) {
            InjectUtils.println("dest dir:" + dest);
            mInjectTool.injectDir(dir);

            List<File> files = new ArrayList<>();
            iteratorPath(dir, files, null);
            for (File file : files) {
                String filePath = file.getAbsolutePath();
                if (filePath.endsWith(".class")
                        && !filePath.contains("R$")
                        && !filePath.contains("R.class")
                        && !filePath.contains("BuildConfig.class")) {
                    mInjectTool.injectFile(dir, file);
                }
            }
        }
    }

    private static void iteratorPath(File dir, List<File> pathName, FileFilter fileFilter) {
        if (!dir.isDirectory()) {
            return;
        }
        File[] files = dir.listFiles(fileFilter);
        if (files != null) {
            for (File file : files) {
                if (file.isFile()) {
                    pathName.add(file);
                } else if (file.isDirectory()) {
                    iteratorPath(file, pathName, fileFilter);
                }
            }
        }
    }
}

File: libplugin/src/main/groovy/com/creation/plugin/inject/ClassTransform.kt
package com.creation.plugin.inject

import com.android.build.api.transform.*
import com.android.build.gradle.TestedExtension
import com.android.build.gradle.internal.pipeline.TransformManager
import com.google.common.collect.Sets
import com.creation.plugin.inject.tool.javaassist.JavaAssistInjectTool

import org.apache.commons.codec.digest.DigestUtils
import org.apache.commons.io.FileUtils
import org.gradle.api.Project

import java.io.File
import java.io.FileFilter
import java.io.IOException
import java.util.ArrayList

internal class ClassTransform(private val mProject: Project, private val mIsApp: Boolean, private val mExtension: TestedExtension) : Transform() {

    private val mInjectTool: IInjectTool

    init {

        mInjectTool = JavaAssistInjectTool(mProject.rootDir.absolutePath)
//        mInjectTool = AsmInjectTool()
    }

    override fun getName(): String {
        return "class editor"
    }

    override fun getInputTypes(): Set<QualifiedContent.ContentType> {
        return TransformManager.CONTENT_CLASS
    }

    override fun getScopes(): MutableSet<in QualifiedContent.Scope> {
        return if (mIsApp) {
            Sets.immutableEnumSet<QualifiedContent.Scope>(
                    QualifiedContent.Scope.PROJECT,
                    QualifiedContent.Scope.SUB_PROJECTS,
                    QualifiedContent.Scope.EXTERNAL_LIBRARIES)
        } else {
            Sets.immutableEnumSet<QualifiedContent.Scope>(
                    QualifiedContent.Scope.PROJECT)
        }
    }

    override fun isIncremental(): Boolean {
        return false
    }

    @Throws(IOException::class)
    override fun transform(context: Context?,
                           inputs: Collection<TransformInput>?,
                           referencedInputs: Collection<TransformInput>?,
                           outputProvider: TransformOutputProvider?,
                           isIncremental: Boolean) {
        InjectUtils.println("transform start")
        mProject.dependencies
        for (input in inputs!!) {
            for (directoryInput in input.directoryInputs) {
                val dest = outputProvider!!.getContentLocation(
                        directoryInput.name,
                        directoryInput.contentTypes,
                        directoryInput.scopes,
                        Format.DIRECTORY)

                try {
                    // 只有这一句是有用的
                    inject(directoryInput.file.absolutePath, dest.absolutePath)
                } catch (e: Exception) {
                    e.printStackTrace()
                }

                FileUtils.copyDirectory(directoryInput.file, dest)
            }

            for (jarInput in input.jarInputs) {
                var jarName = jarInput.name
                val md5Name = DigestUtils.md5Hex(jarInput.file.absolutePath)
                if (jarName.endsWith(".jar")) {
                    jarName = jarName.substring(0, jarName.length - 4)
                }
                val dest = outputProvider!!.getContentLocation(
                        jarName + md5Name,
                        jarInput.contentTypes,
                        jarInput.scopes,
                        Format.JAR)
                FileUtils.copyFile(jarInput.file, dest)
            }
        }
        InjectUtils.println("transform finish")
    }

    private fun inject(path: String, dest: String) {
        val dir = File(path)
        if (dir.isDirectory) {
            InjectUtils.println("dest dir:$dest")
            mInjectTool.injectDir(dir)

            val files = ArrayList<File>()
            iteratorPath(dir, files, null)
            for (file in files) {
                val filePath = file.absolutePath
                if (filePath.endsWith(".class")
                        && !filePath.contains("R$")
                        && !filePath.contains("R.class")
                        && !filePath.contains("BuildConfig.class")) {
                    mInjectTool.injectFile(dir, file)
                }
            }
        }
    }

    private fun iteratorPath(dir: File, pathName: MutableList<File>, fileFilter: FileFilter?) {
        if (!dir.isDirectory) {
            return
        }
        val files = dir.listFiles(fileFilter)
        if (files != null) {
            for (file in files) {
                if (file.isFile) {
                    pathName.add(file)
                } else if (file.isDirectory) {
                    iteratorPath(file, pathName, fileFilter)
                }
            }
        }
    }
}